usaspending_api/data_load/fpds_loader.py

```diff
@@ -0,0 +1,260 @@
+        )
+
+        # matching award. NOT a real db object, but needs to be stored when making the link in load_transactions
+        connected_objects["generated_unique_award_id"] = broker_object["unique_award_key"]
+
+        # award. NOT used if a matching award is found later
+        connected_objects["award"] = _create_load_object(broker_object, None, None, award_functions)
+
+        connected_objects["transaction_normalized"] = _create_load_object(
+            broker_object, transaction_normalized_columns, None, transaction_normalized_functions
+        )
+
+        connected_objects["transaction_fpds"] = _create_load_object(
+            broker_object, transaction_fpds_columns, transaction_fpds_boolean_columns, transaction_fpds_functions
+        )
+
+        retval.append(connected_objects)
+    return retval
+
+
+def _load_transactions(load_objects):
+    """returns ids for each award touched"""
+    ids_of_awards_created_or_updated = set()
+    with psycopg2.connect(dsn=USASPENDING_CONNECTION_STRING) as connection:
+        with connection.cursor(cursor_factory=psycopg2.extras.DictCursor) as cursor:
+
+            # BLIND INSERTS
+            # First create the records that don't have a foreign key out to anything else in one transaction per type
+            inserted_recipient_locations = insert_recipient_locations(cursor, load_objects)
+            for index, elem in enumerate(inserted_recipient_locations):
+                load_objects[index]["legal_entity"]["location_id"] = inserted_recipient_locations[index]
+
+            inserted_recipients = insert_recipients(cursor, load_objects)
+            for index, elem in enumerate(inserted_recipients):
+                load_objects[index]["transaction_normalized"]["recipient_id"] = inserted_recipients[index]
+                load_objects[index]["award"]["recipient_id"] = inserted_recipients[index]
+
+            inserted_place_of_performance = insert_place_of_performance(cursor, load_objects)
+            for index, elem in enumerate(inserted_place_of_performance):
+                load_objects[index]["transaction_normalized"][
+                    "place_of_performance_id"
+                ] = inserted_place_of_performance[index]
+                load_objects[index]["award"]["place_of_performance_id"] = inserted_place_of_performance[index]
+
+            # Handle transaction-to-award relationship for each transaction to be loaded
+            for load_object in load_objects:
+
+                # AWARD GET OR CREATE
+                award_id = _lookup_award_by_transaction(cursor, load_object)
+                if not award_id:
+                    # If there is no award, we need to create one
+                    award_id = insert_award(cursor, load_object)
+
+                load_object["transaction_normalized"]["award_id"] = award_id
+                ids_of_awards_created_or_updated.add(award_id)
+
+                # TRANSACTION UPSERT
```

Comment:
It occurs to me now that this pattern could be made a lot more efficient (2x) by just leveraging Postgres' built-in "UPSERT" functionality: `INSERT ... ON CONFLICT`

https://www.postgresql.org/docs/10/sql-insert.html#SQL-ON-CONFLICT

Example:
```
CREATE TABLE dummy_table (
    id integer PRIMARY KEY,
    val text
);
INSERT INTO dummy_table VALUES(12, 'twelve) ON CONFLICT (id) DO UPDATE SET val=EXCLUDED.val RETURNING id;
INSERT INTO dummy_table VALUES(12, 'new val') ON CONFLICT (id) DO UPDATE SET val=EXCLUDED.val RETURNING id;
```

Using this will eliminate 1 individual `SELECT` _for every_ transaction processed.

This will require altering the sql helper functions, combing the transaction update and insert into e.g. `_upsert_transaction_normalized` and `_upsert_transaction_fpds`

Attempt at what that might look like (note, have to prepend the `EXCLUDED` I believe)
```
upsert_pairs = ["{}=EXCLUDED.{}".format(pair.split('=',1)[0], pair.split('=',1)[1]) for pair in pairs]
sql = "INSERT INTO transaction_fpds {} VALUES {} ON CONFLICT (detached_award_proc_unique) DO UPDATE SET {} RETURNING id".format(columns, values, upsert_pairs)"
```
:warning: Is it correct that `transaction_normalized.transaction_unique_id` is just the broker `detached_award_proc_unique` (for FPDS transactions)?
```
upsert_pairs = ["{}=EXCLUDED.{}".format(pair.split('=',1)[0], pair.split('=',1)[1]) for pair in pairs]
sql = "INSERT INTO transaction_normalized {} VALUES {} ON CONFLICT (transaction_unique_id) DO UPDATE SET {} RETURNING id".format(columns, values, upsert_pairs)"
```

Reply:
NOTE: added _**ONLY DO PART**_ above.